lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/driver/AzureBatchTaskStatusAlarmHandler.java

```diff
@@ -0,0 +1,150 @@
+  }
+
+  /**
+   * This method is periodically invoked by the Runtime Clock. It will call Azure Batch APIs to determine
+   * the status of tasks running inside the job and notify REEF of tasks statuses that correspond to running
+   * evaluators.
+   *
+   * @param alarm alarm object.
+   */
+  @Override
+  public void onNext(final Alarm alarm) {
+    String jobId = this.azureBatchHelper.getAzureBatchJobId();
+    List<CloudTask> allTasks = this.azureBatchHelper.getTaskStatusForJob(jobId);
+
+    // Report status if the task has an associated active container.
+    LOG.log(Level.FINER, "Found {0} tasks from job id {1}", new Object[]{allTasks.size(), jobId});
+    for (CloudTask task : allTasks) {
+      Optional<EvaluatorManager> optionalEvaluatorManager = this.evaluators.get(task.id());
+
+      if (!optionalEvaluatorManager.isPresent() && !TaskState.COMPLETED.equals(task.state())) {
+        // This usually means that the evaluator shim has started, but hasn't sent the status message
+        // back to the driver yet.
+        LOG.log(Level.FINE, "No Evaluator found for Azure Batch task id = {0}. Ignoring.", task.id());
+      } else if (!optionalEvaluatorManager.isPresent() && TaskState.COMPLETED.equals(task.state())) {
+        // This indicates that the evaluator shim exited prematurely. We inform REEF of resource allocation
+        // so it's possible to trigger an event signaling resource failure later.
+        LOG.log(Level.INFO, "Azure Batch task id = {0} is in 'COMPLETED' state, but it does not have " +
+            "an Evaluator associated with it. This indicates that the evaluator shim has failed before " +
+            "it could send a callback to the driver.", task.id());
+        this.evaluatorShimManager.get().onResourceAllocated(task.id(), PLACEHOLDER_REMOTE_ID, Optional.of(task));
+      } else if (optionalEvaluatorManager.get().isClosedOrClosing()) {
+        LOG.log(Level.FINE, "Evaluator id = {0} is closed. Ignoring.", task.id());
+      } else {
+        LOG.log(Level.FINE, "Reporting status for Task Id: {0} is [Azure Batch Status]:{1} ",
+            new Object[]{task.id(), task.state().toString()});
+        this.evaluatorShimManager.get().onAzureBatchTaskStatus(task);
+      }
+    }
+
+    if (this.isAlarmEnabled()) {
+      this.scheduleAlarm();
+    }
+  }
+
+  /**
+   * Enable the period alarm to send status updates.
+   */
+  public synchronized void enableAlarm() {
+    if (!this.isAlarmEnabled) {
+      LOG.log(Level.FINE, "Enabling the alarm and scheduling it to fire in {0} ms.", this.taskStatusCheckPeriod);
+      this.isAlarmEnabled = true;
+      this.scheduleAlarm();
+    } else {
+      LOG.log(Level.FINE, "Alarm is already enabled.");
+    }
+  }
+
+  /**
+   * Disable the period alarm to send status updates.
+   */
+  public synchronized void disableAlarm() {
+    this.isAlarmEnabled = false;
+  }
+
+  private synchronized boolean isAlarmEnabled() {
```

Comment:
In general, I only synchronize public methods in order to avoid race conditions. See comment for line 118 above

Reply:
see comments in line 118.